staging/src/k8s.io/client-go/dynamic/dynamicinformer/informer.go

```diff
@@ -0,0 +1,108 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dynamicinformer
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/dynamic/dynamiclister"
+	"k8s.io/client-go/tools/cache"
+	"sync"
+	"time"
+)
+
+// NewDynamicSharedInformerFactory constructs a new instance of dynamicSharedInformerFactory for all namespaces.
+func NewDynamicSharedInformerFactory(client dynamic.Interface, defaultResync time.Duration) DynamicSharedInformerFactory {
+	return &dynamicSharedInformerFactory{
+		client: client,
+		defaultResync:    defaultResync,
+		informers: map[string]DynamicInformer{},
+	}
+}
+
+type dynamicSharedInformerFactory struct {
+	client    dynamic.Interface
+	defaultResync    time.Duration
+
+	lock             sync.Mutex
+	informers map[string]DynamicInformer
+}
+
+var _ DynamicSharedInformerFactory = &dynamicSharedInformerFactory{}
+
+func (f *dynamicSharedInformerFactory)  ForResource(gvr schema.GroupVersionResource, namespace string) DynamicInformer {
+	f.lock.Lock()
+	defer f.lock.Unlock()
+
+	informer, exists := f.informers[gvr.String()]
+	if exists {
+		return informer
+	}
+
+	// TODO(p0lyn0mial): should we accept cache.Indexer as an argument when constructing the factory ?!
+	informer = NewFilteredDynamicInformer(f.client, gvr, namespace, f.defaultResync, cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc}, nil)
+	f.informers[gvr.String()] = informer
+
+	return informer
+}
+
+// NewFilteredDynamicInformer constructs a new informer for a dynamic type.
+func NewFilteredDynamicInformer(client dynamic.Interface, gvr schema.GroupVersionResource, namespace string, resyncPeriod time.Duration, indexers cache.Indexers, tweakListOptions TweakListOptionsFunc) DynamicInformer {
```

Comment:
Is it practical to write some unit tests for informer ?

Reply:
> Is it practical to write some unit tests for informer ?

I think it should be possible to wire a fake dynamic client in to at least see it call a handler.  If it's not, please show me the broken test so I can convince myself that it's not an issue in the fake dynamic client itself.